### PR TITLE
kinto notifications preamble

### PR DESCRIPTION
**Pull request by phrawzty**
This is the preamble document introducing notifications in Kinto.

It contains Graphviz graphs which are generated in-line by the Graphiviz (Liquid Tags) plugin. This plugin only works in Markdown which is why this document wasn't written in RST.  We also need to ensure that the machine that builds this documentation has Graphviz installed - in particular, `dot`.
- [ ] Validate content.
- [x] Add note re: Graphviz dependency to README.

### CODE REVIEW

**Comment by phrawzty** [comment]
~~Hold for https://github.com/getpelican/pelican-plugins/pull/540 (and subsequent submodule hash update).~~ Good as of 0ab1be0.

**Comment by phrawzty** [comment]
`r?` @ametaireau @leplatrem

File: content/2015.08.notifications-preamble.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+Date: 2015-08-13
+Summary: Notifications are coming to Kinto! Here's the preamble.
+
+At a recent work week in the lovely French town of [Dournenez](https://www.qwant.com/?q=dournenez&t=images) we laid the groundwork for an upcoming feature in [Kinto](https://kinto.readthedocs.org): *Notifications!*
+
+# What?
+
+Before we go into the gritty details, let's take a moment to come to an understanding about some core terminology, shall we? Notifications are messages which trigger actions, the combination of which is called an event. Actions can in turn generate further notifications, resulting in further events, and so forth until the end of time (or thread).
+
+Notifications can come from *somewhere else*, such as an external service, or can be generated internally. Actions can be native functions of Kinto, local executables, features from plugins, and so forth. In general, actions will produce some sort of interesting output. Events can therefore be thought of as a cycle: a message is received, an action takes place, and a message is generated.
```
Comment (almet) [inline content/2015.08.notifications-preamble.md]: "Messages can come from"?

File: content/2015.08.notifications-preamble.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+* The amount of additional code would be relatively small - in essence, only just enough to speak to a queue service and nothing more.
+* We could ship with support for a handful of external queuing services.
+
+Cons:
+
+* We'd have to code and maintain support for a handful of external queuing services.
+* It's ultimately up to the final consumer (receiver, destination, whatever) of the message to deal with the queue. In other words, if the queue is RabbitMQ, then the consumer needs to able to speak RabbitMQ too.
+* Kinto has no control, understanding, or insight into what happens once the message is handed off to the queuing service.
+
+The simplicity of this model, while attractive, has some fatal flaws. Chiefly, it runs contrary to our guideline of tech agnosticism as it would impose a particular external service dependency on the *consumer*. Furthermore, this model flirts with danger in terms of robustness - the fire-and-forget model isn't exactly optimised for reliability.
```
Comment (almet) [inline content/2015.08.notifications-preamble.md]: I am actually not sure this is true: some queuing mechanisms have an ability to check if the message was delivered or not.

File: content/2015.08.notifications-preamble.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+Pros:
+
+* This model implies a highly limited scope insofar as Kinto is directly concerned; we would basically treat Kinto like a lightweight transport mechanism, and the queue as a fire-and-forget target.
+* The amount of additional code would be relatively small - in essence, only just enough to speak to a queue service and nothing more.
+* We could ship with support for a handful of external queuing services.
+
+Cons:
+
+* We'd have to code and maintain support for a handful of external queuing services.
```
Comment (almet) [inline content/2015.08.notifications-preamble.md]: Not sure this is a real problem: if we don't want to do this, then don't implement multiple external queing services (but only one, and support the other ones via plugins, for instance).

File: content/2015.08.notifications-preamble.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+* Similarly, scaling the queue would imply scaling Kinto itself
+
+This model has certain interesting advantages; however, the relative volume of additional code that would need to be developed is reason enough to give us pause ([scope creep](https://en.wikipedia.org/wiki/Scope_creep) is a real problem). Like the previous model, this also runs afoul of our tech agnosticism guideline since we would be locking Kinto to a given queuing service directly. Finally, we have (admittedly theoretical) concerns about how this model would scale - it feels like a waste to increase two resources when only one needs to grow, for example.
+
+## The Broker
+
+The final approach is a bit of a mix of the previous two. In this model, we implement a formal "broker" that is part of the Kinto *eco-system*, but external to Kinto itself. The roker could be anything from a few lines of Python sitting on a [pipe](https://en.wikipedia.org/wiki/Named_pipe) to a proper micro-service stack. This decouples the queuing mechanism from Kinto while preserving a strong *promise* of interoperability and control.
```
Comment (almet) [inline content/2015.08.notifications-preamble.md]: broker.

**Comment by almet** [comment]
I've added some notes. It seems very nice. I was a bit lost at the begining by the introduction of the terms, though.

Also, might be interesting to ask the audience for explicit feedback at the end of the article, and add a tl,dr section at the beginning.

File: content/2015.08.notifications-preamble.md (hunk 3 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md]: This is a counter-point to the _pro_ just above it (i.e. it's a double-edged sword) - anything we commit to supporting means additional work and maintenance down the road.

**Comment by phrawzty** [comment]
I believe that I've addressed the observations, including adding a _tl;dr_ at the start and soliciting feedback at the end. Final `r?` @ametaireau

File: content/2015.08.notifications-preamble.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
+Title: Notifications in Kinto (Preamble)
+Lang: en
+Url: notifications-kinto-preamble
+Slug: notifications-kinto-preamble
+Date: 2015-08-13
+Summary: Notifications are coming to Kinto! Here's the preamble.
+
+At a recent work week in the lovely French town of [Dournenez](https://www.qwant.com/?q=dournenez&t=images) we laid the groundwork for an upcoming feature in [Kinto](https://kinto.readthedocs.org): *Notifications!*
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: Douarnenez

Comment (phrawzty) [inline content/2015.08.notifications-preamble.md]: :eyes:

File: content/2015.08.notifications-preamble.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
+Title: Notifications in Kinto (Preamble)
+Lang: en
+Url: notifications-kinto-preamble
+Slug: notifications-kinto-preamble
+Date: 2015-08-13
+Summary: Notifications are coming to Kinto! Here's the preamble.
+
+At a recent work week in the lovely French town of [Douarnenez](https://www.qwant.com/?q=Douarnenez&t=images) we laid the groundwork for an upcoming feature in [Kinto](https://kinto.readthedocs.org): *Notifications!*
+
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:9]: img_20150806_130242

File: content/2015.08.notifications-preamble.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+# tl;dr
+
+For those that are in a hurry, here's the gist: We're going to be integrating a "notifications" system into Kinto with an eye towards simplicity, modernity, and flexibility. Our goal is to provide a solid framework that will allow other systems to take advantage of this new feature without having to jump through too many hoops. If that sounds interesting to you - read on.
+
+# What?
+
+Before we go into the gritty details, let's take a moment to come to an understanding about some core terminology, shall we? Notifications are messages which trigger actions, the combination of which is called an event. Actions can in turn generate further notifications, resulting in further events, and so forth until the end of time (or thread).
+
+Messages can come from *somewhere else*, such as an external service, or can be generated internally. Actions can be native functions of Kinto, local executables, features from plugins, and so forth. In general, actions will produce some sort of interesting output. Events can therefore be thought of as a cycle: a message is received, an action takes place, and a message is generated.
+
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: I come late in the party, sorry :|

This section could be a lot clearer I believe. Maybe you could split the introduction comment from the definitions.
And maybe put definitions as bullet points ?
- Notifications: ...
- Actions
- Events
- Messages

Otherwise it's bit hard to follow (at least for non native english)

File: content/2015.08.notifications-preamble.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+* We should strive to be as *agnostic* as possible, meaning that locking into a particular service, back-end, or even programming language, should be avoided where possible.
+* Since we're web-centric, we should capitalise on HTTP wherever possible as well.
+* The notifications mechanism must be able to scale with a minimum of hardship; this applies to any eventual queues, workers, and so forth.
+* Finally, it must be robust - we can't be losing messages now, can we?
+
+# How?
+
+There are about as many ways to go about solving the problem as there are grains of sand on a beach (warning: previous statement may not be scientifically valid). The best way to determine the winning approach is simple: through single combat, [Thunderdome-style](https://duckduckgo.com/?q=beyond+thunderdome&ia=about). Unfortunately our manager put the kibosh on that plan for liability reasons, so instead we did the next best thing: we drew on a whiteboard.
+
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:75]: img_20150804_125424

File: content/2015.08.notifications-preamble.md (hunk 3 above)
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: I would reformulate to express the fact that we would not especially be in charge of implementing them.

Something like «every kind of queuing solution would require to code and maintain a dedicated implementation»

File: content/2015.08.notifications-preamble.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            Consumer [shape=box];
+            endpoint -> Consumer [label=4, dir=both];
+        }
+    }
+%}
+Pros:
+
+* Kinto could be shipped as a "turn-key" solution with a fully functioning stack baked right in; while there would be a queuing service involved ([Celery](http://www.celeryproject.org/), for example), it would be completely abstracted by Kinto.
+* Since the queuing mechanism is integral, Kinto would have full visibility and control over processing and delivery.
+* Apart from the queuing service there would be no additional services or software to deploy.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:140]: from the underlying queuing service

File: content/2015.08.notifications-preamble.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+Cons:
+
+* This would involve a fairly serious increase in the scope, complexity, and size of the Kinto code base.
+* The tight coupling means that Kinto, in its entirety, would need to be re-deployed if changes were necessary in the queuing service.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:145]: isn't it : if changes were necessary in consumers code (?)

File: content/2015.08.notifications-preamble.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+* Similarly, scaling the queue would imply scaling Kinto itself
+
+This model has certain interesting advantages; however, the relative volume of additional code that would need to be developed is reason enough to give us pause ([scope creep](https://en.wikipedia.org/wiki/Scope_creep) is a real problem). Like the previous model, this also runs afoul of our tech agnosticism guideline since we would be locking Kinto to a given queuing service directly. Finally, we have (admittedly theoretical) concerns about how this model would scale - it feels like a waste to increase two resources when only one needs to grow, for example.
+
+## The Broker
+
+The final approach is a bit of a mix of the previous two. In this model, we implement a formal "broker" that is part of the Kinto *eco-system*, but external to Kinto itself. The broker could be anything from a few lines of Python sitting on a [pipe](https://en.wikipedia.org/wiki/Named_pipe) to a proper micro-service stack. This decouples the queuing mechanism from Kinto while preserving a strong *promise* of interoperability and control.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:152]: to a proper micro-service stack in Node or Haskell :)

File: content/2015.08.notifications-preamble.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+            Consumer [shape=box];
+            worker -> Consumer [label=4, dir=both];
+        }
+    }
+%}
+
+Pros:
+
+* Provides a clean and easily understandable separation of responsibilities.
+* Frees both Kinto and the consumer from any particular service dependency.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: ...since the transport is HTTP between components

File: content/2015.08.notifications-preamble.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+* Permits the broker to scale independently from Kinto.
+* Requires only a moderate amount of new code.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: ...inside Kinto (the broker is not a piece of cake)

File: content/2015.08.notifications-preamble.md [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+* Effectively allows for a highly flexible solution where just about anything imaginable could be plugged in.
+
+Cons:
+
+* The broker must speak HTTP and respond properly to a specification of our choosing.
+* Functionally requires deploying and maintaining two services (the difficulty of which varies according the complexity of the broker).
+
+We like this model because it satisfies most of our guidelines (and violates none). To be fair, the requirement for "robustness" is a little hand-wavey in this case. We are effectively handing the responsibility to another component; however, since interoperability is a promise, we can maintain an acceptable degree of confidence in the process.
+
+A final point worth mentioning here is that the broker could, in principle, be the *consumer itself*, which fits into some of the potential use-cases at Mozilla quite nicely.
```
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:192]: :+1:

**Comment by leplatrem** [comment]
Very good @phrawzty !

Like alexis I had some trouble understanding the terminology introduction.
But otherwise it is very pleasant to read such summary :)

I haven't tested your graphs locally though.

File: content/2015.08.notifications-preamble.md (hunk 6 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md:9]: notifications-douarnenez

:smile:

File: content/2015.08.notifications-preamble.md (hunk 8 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md:75]: :+1:

File: content/2015.08.notifications-preamble.md (hunk 7 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md]: That's a fair criticism (@ametaireau mentioned something similar). I'll re-work this section entirely.

File: content/2015.08.notifications-preamble.md (hunk 3 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md]: How about "Each queueing service would require a dedicated module to be coded and maintained; this would likely
be an exercise left up to the community as part of the larger ecosystem." :question:

File: content/2015.08.notifications-preamble.md (hunk 9 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md:140]: Why add "underlying" here? It would be the only place in this entire document where that prefix is applied.

File: content/2015.08.notifications-preamble.md (hunk 10 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md:145]: If the queue service changes, then the queuing mechanism must change as well, which (due to the tight coupling) means a re-deploy.

If I'm interpreting what you've said correctly, you have identified a different problem (which I could add to the list, certainly).

File: content/2015.08.notifications-preamble.md (hunk 11 above)
Comment (phrawzty) [inline content/2015.08.notifications-preamble.md:152]: :trollface:

File: content/2015.08.notifications-preamble.md (hunk 3 above)
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md]: :+1:

**Comment by phrawzty** [comment]
The graphs (rendered):

screen shot 2015-08-19 at 12 01 06

screen shot 2015-08-19 at 12 01 13

screen shot 2015-08-19 at 12 01 32

File: content/2015.08.notifications-preamble.md (hunk 6 above)
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:9]: Oh yes please, add some photos to the blogpost :)

File: content/2015.08.notifications-preamble.md (hunk 9 above)
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:140]: Well it was just to distinguish it from the next solution, where the broker has to be deployed. Forget it if it does not bring obvious value :)

File: content/2015.08.notifications-preamble.md (hunk 10 above)
Comment (leplatrem) [inline content/2015.08.notifications-preamble.md:145]: I get it. 
No I don't think consumer are tight in this solution. Forget my comment, sorry.

**Comment by leplatrem** [comment]
I believe to second graph should look like this, in order to constrast it with the third solution, where the broker is external and abstract behind an HTTP POST :)
img_20150819_120937

**Comment by phrawzty** [comment]
I was trying not to use the word "broker" in both of the solutions (hence _internal_queue_ in the second graph). :smile: I'll make an adjustment to my graph in any case as it's a little flawed.

**Comment by phrawzty** [comment]
The rendered graph for the **What?** section and the (adjusted) graph for the **Integrated** model:

screen shot 2015-08-19 at 16 20 59

screen shot 2015-08-19 at 16 21 11

**Comment by phrawzty** [comment]
@leplatrem @ametaireau `r?`

File: content/2015.08.notifications-preamble.md [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    }
+%}
+
+Pros:
+
+* Provides a clean and easily understandable separation of responsibilities.
+* Frees both Kinto and the consumer from any particular service dependency.
+* Permits the broker to scale independently from Kinto.
+* Requires only a moderate amount of new code.
+* Effectively allows for a highly flexible solution where just about anything imaginable could be plugged in.
```
Comment (tarekziade) [inline content/2015.08.notifications-preamble.md:181]: I would also add that the decoupling the broker model simplifies a lot error management when its specific to the use case. e.g. in Balrog use case, dealing with problems like "I nee to try sending again and again the event to Balrog until it works" or "if I get more than 10 write events within an hour, let's merge them into a single event for Balrog"

**Comment by leplatrem** [comment]
a **lot better** thanks !!!

r+ !
